db/mongodb/mongodb.go

```diff
@@ -0,0 +1,221 @@
+package mongodb
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/log"
+)
+
+const (
+	MONGODB_TIMEOUT_CONNECT = 10 * time.Second
+	MONGODB_TIMEOUT_COMMIT  = 10 * time.Second
+	MONGODB_TIMEOUT_QUERY   = 4 * time.Second
+)
+
+type MongoDB struct {
+	db         *mongo.Database
+	collection string
+}
+
+type KeyVal struct {
+	Key   string `bson:"_id" json:"key"` // needs to be an ugly string because of regex
+	Value []byte `bson:"value" json:"value"`
+}
+
+func New(opts db.Options) (*MongoDB, error) {
+	url := os.Getenv("MONGODB_URL")
+
+	if url == "" {
+		return nil, fmt.Errorf("missing MONGODB_URL env var")
+	}
+	// we use the path as the collection name (and we hash it to avoid invalid chars)
+	collection := fmt.Sprintf("%x", sha256.Sum256([]byte(opts.Path)))[:12]
+	log.Debugw("connecting to mongo database", "url", url, "collection", collection)
+	clientOptions := options.Client().ApplyURI(url).SetMaxConnecting(20)
+	ctx, cancel := context.WithTimeout(context.Background(), MONGODB_TIMEOUT_CONNECT)
+	defer cancel()
+	client, err := mongo.Connect(ctx, clientOptions)
+	if err != nil {
+		return nil, err
+	}
+	database := client.Database(collection)
+
+	// Shutdown database connection when SIGTERM received
+	go func() {
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+		<-c
+		log.Warnf("received SIGTERM, disconnecting mongo database")
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		err := client.Disconnect(ctx)
+		if err != nil {
+			log.Warn(err)
+		}
+		cancel()
+	}()
+
+	err = client.Ping(ctx, readpref.Primary())
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to mongodb: %w", err)
+	}
+
+	return &MongoDB{db: database, collection: collection}, nil
+}
+
+func (d *MongoDB) Close() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	return d.db.Client().Disconnect(ctx)
+}
+
+func (d *MongoDB) WriteTx() db.WriteTx {
+	return &WriteTx{
+		db:         d.db,
+		batch:      make([]mongo.WriteModel, 0),
+		inMem:      make(map[string][]byte),
```

Comment:
you route all Get calls through WriteTx, which is fine since it's only a wrapper, but then this allocation is always a waste for any get call. I would keep `inMem` nil initially, which is fine because any read operation on a nil map works the same as a non-nil empty map. you only need to check if the map is nil, and make it, before you insert an entry in Set.